### PR TITLE
Add DirectCLR (#781)

### PR DESCRIPTION
**Pull request by KylevdLangemheen**
Issue #781 proposes adding DirectCLR to Lightly. In PR #963 a loss was proposed but never merged.

As the paper uses a version of [infoNCE that is identical to NTX-ent](https://discordapp.com/channels/@me/1399964647523876874/1400461672883159160), we can use our existing implementation of NTX-ent instead. This way, implementing DirectCLR becomes a simple matter of taking the first _d_ values of the embedding and applying the NTX-ent loss to that instead (if my understanding is correct :smile:  ).

As DirectCLR does away with a projection head, I did not see an obvious way to implement it under low level module blocks. The proposed implementation is added as a deprecated model, but perhaps it does not even need to get its own model file and can just live as an example.

This PR:
- Adds DirectCLR as deprecated model
- Adds example for DirectCLR
- Adds test for DirectCLR

This PR _does not_:
- Add DirectCLR to lightly cli
- Implement benchmarks for DirectCLR

Let me know if you would prefer to see a variant outside of the deprecated high-level models. I could make a "dummy" projection head which just slices the embedding for example.

### CODE REVIEW

File: examples/pytorch/directclr.py [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+
+import torch
+import torchvision
+from torch import nn
+
+from lightly.loss import NTXentLoss
+from lightly.models import DirectCLR
+from lightly.transforms.simclr_transform import SimCLRTransform
+
+resnet = torchvision.models.resnet18()
```
[the remark this answers is not in the document]
Comment (KylevdLangemheen) [reply examples/pytorch/directclr.py:16]: Changed it, also for `torchvision.datasets`

File: tests/models/test_ModelsDirectCLR.py [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+    backbone = nn.Sequential(
+        lightly.models.batchnorm.get_norm_layer(3, 0),
+        *list(resnet.children())[:-1],
+        nn.Conv2d(last_conv_channels, num_ftrs, 1),
+        nn.AdaptiveAvgPool2d(1),
+    )
+    return backbone
+
+
+class TestModelsDirectCLR(unittest.TestCase):
```
[the remark this answers is not in the document]
Comment (KylevdLangemheen) [reply tests/models/test_ModelsDirectCLR.py]: In the current set-up I've made DirectCLR loss inherit from the NTXent loss. As such, many tests become redundant. I currently test whether it can take various loss_dims, whether some of the calculations (that use NTXent) are done correctly, and whether the parent parameter (temperature) is set correctly. Let me know if some tests are missing or redundant still!

File: lightly/models/directclr.py [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+
+        # return out0 if x1 is None
+        if x1 is None:
+            # return features if requested
+            if return_features:
+                return out0, f0
+            return out0
+
+        # forward pass of second input x1
+        f1: Tensor = self.backbone(x1).flatten(start_dim=1)
```
[the remark this answers is not in the document]
Comment (KylevdLangemheen) [reply lightly/models/directclr.py]: Now that it is no longer a model but an example, I have included this with the example. Let me know if it is better to keep it simple in the example instead.

File: examples/pytorch/directclr.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (KylevdLangemheen) [reply examples/pytorch/directclr.py]: I don´t know exactly where the line is drawn (i.e. nn.Sequential). There's some variance in the codebase I think?

[the remark this answers is not in the document]
Comment (liopeer) [reply examples/pytorch/directclr.py]: It's actually a hard line haha. :) Unfortunately a lot of the code is not up-to-date.

File: lightly/loss/directclr_loss.py [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+    head. Set loss_dim to the desired truncated representation length.
+    DirectCLRLoss inherits from NTXentLoss, its parameters can be set after
+    setting loss_dim.
+
+    - [0] DirectCLR, 2021, https://arxiv.org/abs/2110.09348
+
+    Attributes:
+        loss_dim:
+            Computes the loss only on the first loss_dim values of the encoding.
+        *args:
```
Comment (liopeer) [inline lightly/loss/directclr_loss.py:27]: Since we have full control over the `NTXentLoss` and its arguments it's usually preferred to be verbose and copy the arguments from the NTXentLoss over here. I would only use `*args` and `**kwargs` when the subsequently instantiated class comes from a third-party package and the interface might change unexpectedly – and even in that case I would only allow passing keyword arguments and not positional arguments.

File: examples/pytorch/directclr.py [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+from torch.optim import SGD
+from torch.utils.data import DataLoader
+from torchvision import models
+from torchvision.datasets import CIFAR10
+
+from lightly.loss import DirectCLRLoss
+from lightly.transforms.simclr_transform import SimCLRTransform
+
+resnet = models.resnet18()
+model = nn.Sequential(*list(resnet.children())[:-1])
```
Comment (liopeer) [inline examples/pytorch/directclr.py:19]: Nit:
```suggestion
model = Sequential(*list(resnet.children())[:-1])
```

File: lightly/loss/directclr_loss.py [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+from torch import Tensor
+
+from lightly.loss import NTXentLoss
+
+
+class DirectCLRLoss(NTXentLoss):
+    """Implementation of the NT-Xent based DirectCLR Loss.
+
+    Following the DirectCRL[0] paper, this loss should be used without projection
+    head. Set loss_dim to the desired truncated representation length.
```
Comment (liopeer) [inline lightly/loss/directclr_loss.py:18]: Nit:
```suggestion
    head. Set `loss_dim` to the desired truncated representation length.
```
This will render nicely once we add it to the docs. Same thing further down.

File: lightly/loss/directclr_loss.py [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+                Shape: (batch_size, embedding_size)
+            out1:
+                Output projections of the second set of transformed images.
+                Shape: (batch_size, embedding_size)
+
+        Returns:
+            DirectCLR Loss value.
+        """
+
+        out0 = out0.flatten(start_dim=1)[:, : self.loss_dim]
```
Comment (liopeer) [inline lightly/loss/directclr_loss.py:86]: Flattening would not really be necessary since you have a really nice docstring that tells the user how it should be used. But you can leave it as is.

File: tests/loss/test_directclr_loss.py [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+    @pytest.mark.parametrize("n_samples", [1, 2, 4])
+    @pytest.mark.parametrize("dimension", [1, 2, 16, 64])
+    @pytest.mark.parametrize("loss_dim", [1, 2, 32])
+    def test_with_values(
+        self,
+        n_samples: int,
+        dimension: int,
+        loss_dim: int,
+    ) -> None:
+        out0 = torch.tensor(
```
Comment (liopeer) [inline tests/loss/test_directclr_loss.py:25]: You could also directly use `torch.randn`, but it's of course totally okay this way as well.

File: tests/loss/test_directclr_loss.py [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+
+class TestDirectCLRLoss:
+    def test_set_parent_temperature(self) -> None:
+        loss_function = DirectCLRLoss(loss_dim=32, temperature=0.5)
+        assert (
+            loss_function.temperature == 0.5
+        ), f"Expected temperature to be 0.5, but got {loss_function.temperature}"
+
+    @pytest.mark.parametrize("n_samples", [1, 2, 4])
+    @pytest.mark.parametrize("dimension", [1, 2, 16, 64])
```
Comment (liopeer) [inline tests/loss/test_directclr_loss.py:17]: I would use smaller dimensions here. We have a lot of unit tests in LightlySSL, so we should always try to use the fastest possible unit tests – ofc without compromising correctness.

File: tests/loss/test_directclr_loss.py [hunk 10]
```diff
@@ -0,0 +1,107 @@
...
+        l2_manual = _calc_directclr_loss_manual(
+            out0, out1, loss_dim=loss_dim, temperature=0.5
+        )
+        assert l1 == pytest.approx(l2, abs=1e-5)
+        assert l1 == pytest.approx(l1_manual, abs=1e-5)
+        assert l2 == pytest.approx(l2_manual, abs=1e-5)
+
+    def test_forward_pass(self) -> None:
+        loss = DirectCLRLoss(loss_dim=32, temperature=0.5, memory_bank_size=0)
+        for bsz in range(1, 20):
```
Comment (liopeer) [inline tests/loss/test_directclr_loss.py:47]: Also here we should speed things up a notch. I would test max 2 cases: Batch size 1 and something else. And I would parametrize the test, so that you get a dedicated test result for each case. Parametrizing would also allow to use e.g. `pytest-xdist` to run the tests in parallel.

File: tests/loss/test_directclr_loss.py [hunk 11]
```diff
@@ -0,0 +1,107 @@
...
+            batch_2 = torch.randn((bsz, 64))
+
+            # symmetry
+            l1 = loss(batch_1, batch_2)
+            l2 = loss(batch_2, batch_1)
+            assert (l1 - l2).pow(2).item() == pytest.approx(0.0)
+
+    def test_forward_pass_1d(self) -> None:
+        loss = DirectCLRLoss(loss_dim=32, temperature=0.5, memory_bank_size=0)
+        for bsz in range(1, 20):
```
Comment (liopeer) [inline tests/loss/test_directclr_loss.py:58]: Same thing here.

**Review by liopeer — CHANGES_REQUESTED** [review]
Awesome work! It's looking really nice and many of my comments are really just nitpicking and beautification.

The two things we should really change however before merging: The `*args, **kwargs` story, and the *slightly too comprehensive* 🙃 unit tests. After that we're ready to merge.

**Comment by yutong-xiang-97** [comment]
Quick note: the `examples/` are for the supported models / SSL methods in the docs (like shown [here](https://docs.lightly.ai/self-supervised-learning/examples/simclr.html)) but not for the losses. We could have a follow-up issue implementing the whole DirectCLR at `benchmarks/`. Then we can add examples and docs for it.

**Review by yutong-xiang-97 — COMMENTED** [review]
Added my review as well.

**Comment by KylevdLangemheen** [comment]
I have:
- Generated the example notebooks with `make generate-example-notebooks`
- Reduced dims in tests
- Fixed typo
- Added formatting for loss_dim
- Directly import functions/classes at the module level
- Use `torch.randn` consistently
- Added NTXentLoss arguments directly

Is there a standard way to ensure or even automate that the parameters relating to NTXentLoss always match? This way the defaults can get out of sync pretty easily :(

**Comment by KylevdLangemheen** [comment]
Resolved the temperature parameter and default parameter comments.

**Review by liopeer — APPROVED** [review]
Everything addressed from my side as well. Amazing work @KylevdLangemheen !
